Make kubectl bash-completion namespace and resource alias aware

sttts opened the pull request:
![kubectl-completion](https://cloud.githubusercontent.com/assets/730123/13912456/94abf9a2-ef3d-11e5-9c8a-3df7058ea8f0.gif)
- filter resource listing by `--namespace` flag given before in the command line

``` bash
$ kubectl get pod --namespace=kube-system <tab><tab>
kube-dns-v9-2wuzj   kube-dns-v9-llqxa
```
- add completion of `--namespace`

``` bash
$ kubectl get pod --namespace=<tab><tab>
[*]         default      ingress      kube-system
```
- add support for plural nouns and aliases like `rc`

sttts committed:
Update cobra godep

sttts committed:
Make kubectl bash completion namespace aware and add noun aliases

- add namespace filtering to bash completion
- add --namespace flag bash completion
- add bash completion noun aliases
- adapt to new cobra package structure

sttts committed:
Update generated contrib/completions/bash/kubectl

eparis commented on pkg/kubectl/cmd/cmd.go:
```diff
@@ -30,12 +30,47 @@ import (
+    local ret two_word_ns
+    ret=""
+    two_word_ns=false
+    for w in ${words[@]}; do
```
${words[@]} should be in quotes.

sttts replied:
Fixed.

eparis commented on contrib/completions/bash/kubectl:
```diff
@@ -135,6 +166,8 @@ __handle_noun()
 
     if __contains_word "${words[c]}" "${must_have_one_noun[@]}"; then
         must_have_one_noun=()
+    elif __contains_word "${words[c]%!s(MISSING)}" "${must_have_one_noun[@]}"; then
```
`%!s(MISSING)` ?

sttts replied:
Bad escaping of bash code in golang string in fact: https://github.com/spf13/cobra/pull/255/files#diff-3db5cab63dbb32fff4264f5449fd2904R192

sttts commented:
hack/update-godep-licenses.sh doesn't like me.

sttts commented:
bash4'ism in that script.

sttts commented:
ptal @eparis. Everything green.

sttts commented:
- namespaced all zsh function to avoid conflicts with git completion and others
- updated godeps to https://github.com/spf13/cobra/pull/257
- added noun aliases
- added proper compgen filtering for zsh

sttts commented:
Rebased. But because it's based on unmerged spf13/cobra#257, it's still red due to invalid godeps.

eparis commented on contrib/completions/zsh/kubectl:
```diff
@@ -0,0 +1,214 @@
+#!/bin/zsh
```
This file was just created my hand right? Not generated in any way? Should it be a separate PR?

sttts replied:
Yes, it's just a collection of functions that zsh's bash-completion emulation does not provide or which are incomplete. Will create an independent PR.

sttts replied:
Moved into https://github.com/kubernetes/kubernetes/pull/23797

sttts commented:
Updated to latest cobra master with shellcheck fixes.

sttts commented:
@k8s-bot test this please flake: #21367
